File: backend/src/PMP.Infrastructure/Services/Chat/ChatService.cs
using Microsoft.AspNetCore.SignalR;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.EntityFrameworkCore;
using PMP.Infrastructure.Services.System;
using PMP.Infrastructure.Hubs;
using PMP.Application.Features.Chat.DTOs;
using PMP.Application.Features.Chat.Interfaces;
using PMP.Domain.Entities.Chat;
using PMP.Domain.Enums;
using PMP.Infrastructure.Persistence;
using PMP.Shared.Wrappers;

namespace PMP.Infrastructure.Services.Chat;

public class ChatService : IChatService
{
    private readonly ApplicationDbContext _db;
    private readonly IHubContext<ChatHub> _hubContext;
    private readonly IServiceScopeFactory _scopeFactory;

    public ChatService(ApplicationDbContext db, IHubContext<ChatHub> hubContext, IServiceScopeFactory scopeFactory)
    {
        _db = db;
        _hubContext = hubContext;
        _scopeFactory = scopeFactory;
    }

    // ─── Conversations ───────────────────────────────────────────────────────

    public async Task<ApiResponse<List<ConversationDto>>> GetUserConversationsAsync(Guid userId)
    {
        var convs = await _db.Conversations
            .Where(c => c.UserId == userId && c.IsActive)
            .OrderByDescending(c => c.LastMessageAt ?? c.CreatedAt)
            .Select(c => new ConversationDto
            {
                Id = c.Id,
                Title = c.Title,
                Type = (int)c.Type,
                LastMessageAt = c.LastMessageAt,
                LastMessage = c.Messages.OrderByDescending(m => m.CreatedAt).Select(m => new MessageDto
                {
                    Content = m.Content,
                    CreatedAt = m.CreatedAt
                }).FirstOrDefault()
            })
            .ToListAsync();

        return new ApiResponse<List<ConversationDto>>(convs);
    }

    public async Task<ApiResponse<ConversationDto>> CreateConversationAsync(Guid userId, CreateConversationRequest req)
    {
        var conv = new Conversation
        {
            UserId = userId,
            Type = (ConversationType)req.Type,
            Title = string.IsNullOrEmpty(req.InitialMessage) 
                ? "Cuộc hội thoại mới" 
                : (req.InitialMessage.Length > 30 ? req.InitialMessage[..30] + "..." : req.InitialMessage),
            IsActive = true,
            LastMessageAt = DateTime.UtcNow
        };

        _db.Conversations.Add(conv);
        await _db.SaveChangesAsync();

        if (!string.IsNullOrEmpty(req.InitialMessage))
        {
            var msg = new Message
            {
                ConversationId = conv.Id,
                Role = MessageRole.User,
                Content = req.InitialMessage,
                ContentType = MessageContentType.Text
            };
            _db.Messages.Add(msg);
            await _db.SaveChangesAsync();
        }

        return new ApiResponse<ConversationDto>(new ConversationDto
        {
            Id = conv.Id,
            Title = conv.Title,
            Type = (int)conv.Type,
            LastMessageAt = conv.LastMessageAt
        });
    }

    public async Task<ApiResponse<bool>> DeleteConversationAsync(Guid userId, Guid convId)
    {
        var conv = await _db.Conversations.FirstOrDefaultAsync(c => c.Id == convId && c.UserId == userId);
        if (conv == null) return new ApiResponse<bool>("Không tìm thấy cuộc hội thoại.");

        conv.IsActive = false; // Soft delete
        await _db.SaveChangesAsync();
        return new ApiResponse<bool>(true, "Đã xoá cuộc hội thoại.");
    }

    // ─── Messages ────────────────────────────────────────────────────────────

    public async Task<ApiResponse<List<MessageDto>>> GetConversationMessagesAsync(Guid userId, Guid convId)
    {
        var conv = await _db.Conversations.AnyAsync(c => c.Id == convId && c.UserId == userId);
        if (!conv) return new ApiResponse<List<MessageDto>>("Không tìm thấy cuộc hội thoại.");

        var msgs = await _db.Messages
            .Where(m => m.ConversationId == convId)
            .OrderBy(m => m.CreatedAt)
            .Select(m => MapMessage(m))
            .ToListAsync();

        return new ApiResponse<List<MessageDto>>(msgs);
    }

    public async Task<ApiResponse<MessageDto>> SendMessageAsync(Guid userId, SendMessageRequest req)
    {
        var conv = await _db.Conversations.FirstOrDefaultAsync(c => c.Id == req.ConversationId && c.UserId == userId);
        if (conv == null) return new ApiResponse<MessageDto>("Không tìm thấy cuộc hội thoại.");

        var msg = new Message
        {
            ConversationId = req.ConversationId,
            Role = MessageRole.User,
            Content = req.Content,
            AttachmentUrl = req.AttachmentUrl,
            ContentType = MessageContentType.Text
        };

        _db.Messages.Add(msg);
        conv.LastMessageAt = DateTime.UtcNow;
        await _db.SaveChangesAsync();

        var dto = MapMessage(msg);

        // Broadcast to group (real-time update for other client tabs or admin)
        await _hubContext.Clients.Group(req.ConversationId.ToString()).SendAsync("ReceiveMessage", dto);

        // If it's an AI conversation, we could trigger AI response here
        if (conv.Type == ConversationType.AI)
        {
            _ = Task.Run(() => HandleAiResponse(userId, conv.Id, req.Content));
        }

        return new ApiResponse<MessageDto>(dto);
    }

    // ─── AI Response Simulation ──────────────────────────────────────────────

    private async Task HandleAiResponse(Guid userId, Guid convId, string userMessage)
    {
        using var scope = _scopeFactory.CreateScope();
        var aiChatService = scope.ServiceProvider.GetRequiredService<IAiChatService>();
        var db = scope.ServiceProvider.GetRequiredService<ApplicationDbContext>();

        var aiResponse = await aiChatService.GetAiResponseAsync(userId, userMessage);
        var content = aiResponse.Succeeded ? aiResponse.Data! : (aiResponse.Message ?? "Xin lỗi, tôi gặp sự cố khi kết nối với não bộ AI.");

        var aiMsg = new Message
        {
            ConversationId = convId,
            Role = MessageRole.Assistant,
            Content = content,
            ContentType = MessageContentType.Text,
            CreatedAt = DateTime.UtcNow
        };

        db.Messages.Add(aiMsg);
        await db.SaveChangesAsync();

        var dto = MapMessage(aiMsg);
        await _hubContext.Clients.Group(convId.ToString()).SendAsync("ReceiveMessage", dto);
    }

    private static MessageDto MapMessage(Message m) => new()
    {
        Id = m.Id,
        ConversationId = m.ConversationId,
        Role = (int)m.Role,
        Content = m.Content,
        ContentType = (int)m.ContentType,
        AttachmentUrl = m.AttachmentUrl,
        IsStreaming = m.IsStreaming,
        CreatedAt = m.CreatedAt
    };
}

File: frontend/src/services/chat/chatService.ts
import apiClient from '@/services/apiClient';
import * as signalR from '@microsoft/signalr';
import { CONFIG } from '@/config';

// ─── Types ────────────────────────────────────────────────────────────────────

export interface Message {
  id: string;
  conversationId: string;
  role: number;           // 0=User, 1=Assistant, 2=System
  content: string;
  contentType: number;    // 0=Text, 1=Markdown, 2=Image
  attachmentUrl?: string;
  isStreaming: boolean;
  createdAt: string;
}

export interface Conversation {
  id: string;
  title?: string;
  type: number;           // 0=AI, 1=Admin
  lastMessageAt?: string;
  lastMessage?: Partial<Message>;
}

// ─── API ─────────────────────────────────────────────────────────────────────

export const chatService = {
  getConversations: async (): Promise<Conversation[]> => {
    const res = await apiClient.get('/chat/conversations');
    return res.data.data;
  },
  createConversation: async (data: { type: number; initialMessage?: string }): Promise<Conversation> => {
    const res = await apiClient.post('/chat/conversations', data);
    return res.data.data;
  },
  deleteConversation: async (id: string) => apiClient.delete(`/chat/conversations/${id}`),

  getMessages: async (conversationId: string): Promise<Message[]> => {
    const res = await apiClient.get(`/chat/conversations/${conversationId}/messages`);
    return res.data.data;
  },
  sendMessage: async (data: { conversationId: string; content: string; attachmentUrl?: string }): Promise<Message> => {
    const res = await apiClient.post('/chat/messages', data);
    return res.data.data;
  },
};

// ─── SignalR Helper ───────────────────────────────────────────────────────────

let hubConnection: signalR.HubConnection | null = null;
let currentJoinedConversationId: string | null = null;

export const chatHub = {
  start: async (token: string, onMessageReceived: (msg: Message) => void) => {
    if (hubConnection) {
      hubConnection.off('ReceiveMessage');
      hubConnection.on('ReceiveMessage', onMessageReceived);
      if (hubConnection.state === signalR.HubConnectionState.Connected && currentJoinedConversationId) {
        try {
          await hubConnection.invoke('JoinConversation', currentJoinedConversationId);
        } catch (e) {
          console.error('Error invoking JoinConversation on start:', e);
        }
      }
      return;
    }

    const baseUrl = CONFIG.API_BASE_URL.replace(/\/api$/, '');
    const hubUrl = `${baseUrl}/hubs/chat`;

    hubConnection = new signalR.HubConnectionBuilder()
      .withUrl(hubUrl, {
        accessTokenFactory: () => token
      })
      .withAutomaticReconnect()
      .build();

    hubConnection.on('ReceiveMessage', (message: Message) => {
      onMessageReceived(message);
    });

    try {
      await hubConnection.start();
      console.log('SignalR Connected.');
      if (currentJoinedConversationId) {
        await hubConnection.invoke('JoinConversation', currentJoinedConversationId);
      }
    } catch (err) {
      console.error('SignalR Connection Error: ', err);
    }
  },
  stop: async () => {
    if (hubConnection) {
      await hubConnection.stop();
      hubConnection = null;
    }
  },
  joinConversation: async (id: string) => {
    currentJoinedConversationId = id;
    if (hubConnection?.state === signalR.HubConnectionState.Connected) {
      await hubConnection.invoke('JoinConversation', id);
    }
  },
  leaveConversation: async (id: string) => {
    if (currentJoinedConversationId === id) {
      currentJoinedConversationId = null;
    }
    if (hubConnection?.state === signalR.HubConnectionState.Connected) {
      await hubConnection.invoke('LeaveConversation', id);
    }
  }
};
